### PR TITLE
TypeIntrospection example implementation [14797]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

Implement a new "generic" Type Introspection example.

## Description

This new example added in `examples/C++/DDS` is able to:
- Create a DomainParticipant with a Publisher and a DataWriter with different types depending on the arguments of the process. These datatypes are different from each other. These are :
  - Created by FastDDSGen and with TypeObject information
  - Created dynamically by an xml
  - Created dynamically by DynamicTypes
- Create a DomainParticipant that, when discover and endpoint in the topic specificed (by argument) it creats a Subscriber and a DataReader in that topic and with the data type registered dynamically. This Reader is able to decode the message and print it in `stdout` with the name of each variable and its value.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [ ] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [ ] Documentation builds and tests pass locally. 
- [ ] New feature has been added to the `versions.md` file (if applicable).
- [ ] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by jparisu** [comment]
@richiprosima Please test this

File: examples/cpp/dds/TypeIntrospectionExample/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+# Basic Configuration Example
```
Comment (irenebm) [inline examples/cpp/dds/TypeIntrospectionExample/README.md:1]:
```suggestion
# Basic Configuration Example
# Type Introspection Example
```

File: examples/cpp/dds/TypeIntrospectionExample/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        message(FATAL_ERROR "Compiler doesn't support C++11")
+    endif()
+endif()
+
+# Get executable sources
+message(STATUS "Configuring TypeIntrospection example...")
+
+file(GLOB IMPLEMENTATION_SOURCES_CPP "*.cpp")
+file(GLOB DATATYPEBASE_SOURCES_CPP "types/*.cpp")
+# file(GLOB DATATYPE_HELLOWORLD_SOURCES_CPP "types/hello_world/**/*.cpp")
```
Comment (irenebm) [inline examples/cpp/dds/TypeIntrospectionExample/CMakeLists.txt:48]: Remove commented lines?

File: examples/cpp/dds/TypeIntrospectionExample/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+General options:
+  -h              --help
+                    Produce help message.
+
+Publisher options:
+  -t <topic_name> --topic=<topic_name>
+                    Topic name (Default: HelloWorldTopic).
+  -d <id>         --domain=<id>
+                    DDS domain ID (Default: 0).
+  -w <num>        --wait=<num>
```
Comment (irenebm) [inline examples/cpp/dds/TypeIntrospectionExample/README.md:41]: Remove

File: examples/cpp/dds/TypeIntrospectionExample/README.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+### Publisher Data Type
+
+The Publisher has an argument `type`.
+Depending on this argument, the Publisher will create a DataWriter using different Data Types.
+Each of the types available are registered in the Participant in different ways, to illustrate how this could
+be done.
+
+This Data Types are:
+
+TODO
```
Comment (irenebm) [inline examples/cpp/dds/TypeIntrospectionExample/README.md:87]: TODO

File: examples/cpp/dds/TypeIntrospectionExample/arg_configuration.h [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+#include <optionparser.hpp>
+
+namespace option = eprosima::option;
+
+constexpr const char* HELLO_WORLD_DATA_TYPE_ARG = "hw";
+constexpr const char* ARRAY_DATA_TYPE_ARG = "array";
+constexpr const char* SEQUENCE_DATA_TYPE_ARG = "seq";
+constexpr const char* STRUCT_DATA_TYPE_ARG = "struct";
+constexpr const char* PLAIN_DATA_TYPE_ARG = "plain";
+constexpr const char* SIMPLELARGE_DATA_TYPE_ARG = "simple";
```
Comment (jparisu) [inline examples/cpp/dds/TypeIntrospectionExample/arg_configuration.h:35]: Let`s use consistent names and call them equal to their folders 
```suggestion
constexpr const char* SIMPLELARGE_DATA_TYPE_ARG = "simple_large";
```

**Review by jparisu — COMMENTED** [review]
Changes requested

**Comment by JLBuenoLopez-eProsima** [comment]
Milestone updated to v2.11.0 because the example fails with current dynamic types implementation.

File: examples/cpp/dds/TypeIntrospectionExample/arg_configuration.h [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+    },
+    { DOMAIN_ID, 0, "d", "domain",                Arg::Numeric,
+      "  -d <id> \t--domain=<id>  \tDDS domain ID (Default: 0)." },
+    { SAMPLES, 0, "s", "samples",              Arg::Numeric,
+      "  -s <num> \t--samples=<num>  \tNumber of samples to send (Default: 0 => infinite samples)." },
+    { TYPE_OBJECT, 0, "o", "type-object",              Arg::None,
+      "  -o \t--type-object=  \tUse Type Object to send Data Type Introspection info." },
+    { TYPE_INFORMATION, 1, "i", "type-information",              Arg::None,
+      "  -i \t--type-information=  \tUse Type Information to send Data Type Introspection info." },
+    { TYPE_INFORMATION_DISABLE, 0, "i", "type-information-disable",              Arg::None,
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/TypeIntrospectionExample/arg_configuration.h:231]: Please check this works as expected: verify if enumeration should coincide and/or second argument should be 0 or 1. Also, perhaps one of them should not use `-i`.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: examples/cpp/dds/TypeIntrospectionExample/TypeIntrospectionSubscriber.cpp [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            participant_->delete_topic(topic_);
+        }
+        if (subscriber_ != nullptr)
+        {
+            if (reader_ != nullptr)
+            {
+                subscriber_->delete_datareader(reader_);
+            }
+            participant_->delete_subscriber(subscriber_);
+        }
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/TypeIntrospectionExample/TypeIntrospectionSubscriber.cpp:142]:
```suggestion
        if (subscriber_ != nullptr)
        {
            if (reader_ != nullptr)
            {
                subscriber_->delete_datareader(reader_);
            }
            participant_->delete_subscriber(subscriber_);
        }
        if (topic_ != nullptr)
        {
            participant_->delete_topic(topic_);
        }
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Pull request closed without merging**